polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -703,13 +710,218 @@ void AffineParallelGuardBuilder::getConstraints(
+    NamedAttrList attrs;
+    attrs.set(mlir::sycl::SYCLDialect::getArgumentTypesAttrName(),
+              builder.getTypeArrayAttr({range.getType(), indexOp.getType()}));
+    attrs.set(mlir::sycl::SYCLDialect::getFunctionNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("get")));
+    attrs.set(mlir::sycl::SYCLDialect::getTypeNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("range")));
+    const auto resTy = builder.getI64Type();
+    return builder.create<sycl::SYCLRangeGetOp>(
+        loc, resTy, ValueRange({range, indexOp}), attrs);
+  }
+
+  static sycl::SYCLAccessorGetRangeOp
+  createSYCLAccessorGetRangeOp(TypedValue<MemRefType> accessor,
+                               OpBuilder builder, Location loc) {
+    const auto accTy =
+        cast<sycl::AccessorType>(accessor.getType().getElementType());
+    const auto rangeTy = cast<sycl::RangeType>(
+        cast<sycl::AccessorImplDeviceType>(accTy.getBody()[0]).getBody()[1]);
+    NamedAttrList attrs;
+    attrs.set(mlir::sycl::SYCLDialect::getArgumentTypesAttrName(),
+              builder.getTypeArrayAttr(accessor.getType()));
+    attrs.set(mlir::sycl::SYCLDialect::getFunctionNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("get_range")));
+    attrs.set(mlir::sycl::SYCLDialect::getTypeNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("accessor")));
+    return builder.create<sycl::SYCLAccessorGetRangeOp>(loc, rangeTy, accessor,
+                                                        attrs);
+  }
+
+  static sycl::SYCLAccessorSubscriptOp
+  createSYCLAccessorSubscriptOp(TypedValue<MemRefType> accessor,
+                                TypedValue<MemRefType> id, OpBuilder builder,
+                                Location loc) {
+    const auto accTy =
+        cast<sycl::AccessorType>(accessor.getType().getElementType());
+    const auto MT = cast<MemRefType>(
+        cast<LLVM::LLVMStructType>(accTy.getBody()[1]).getBody()[0]);
+    NamedAttrList attrs;
+    attrs.set(mlir::sycl::SYCLDialect::getArgumentTypesAttrName(),
+              builder.getTypeArrayAttr({accessor.getType(), id.getType()}));
+    attrs.set(mlir::sycl::SYCLDialect::getFunctionNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("operator[]")));
+    attrs.set(mlir::sycl::SYCLDialect::getTypeNameAttrName(),
+              FlatSymbolRefAttr::get(builder.getStringAttr("accessor")));
+    return builder.create<sycl::SYCLAccessorSubscriptOp>(
+        loc, MT, ValueRange({accessor, id}), attrs);
+  }
+
+  static sycl::SYCLAccessorSubscriptOp
+  getSYCLAccessorBegin(TypedValue<MemRefType> accessor, OpBuilder builder,
+                       Location loc) {
+    const auto accTy =
+        cast<sycl::AccessorType>(accessor.getType().getElementType());
+    const auto idTy = cast<sycl::IDType>(
+        cast<sycl::AccessorImplDeviceType>(accTy.getBody()[0]).getBody()[0]);
```

Comment:
NIT: Instead of relying on the body of the type, we could be creating a new `IDType` here... It's also true we'd need to be creating the body in an ad-hoc way, so I don't have a strong preference.

Reply:
Keeping to ensure we are creating the same IDType as the one used in AccessorType.